### DISCUSSION TITLE
Is it possible to send and receive services?

### DISCUSSION THREAD

**Discussion by alexgiano** [ika-rwth-aachen/mqtt_client#70] [General]
Is it possible to send services via RPC calls? I saw that it is possible to exploit libraries like the following [https://github.com/ika-rwth-aachen/mqtt_client](url) to send asynchronous messages and wait for a response just like in ROS. For my project it is essential to use services to send commands to my robot. Thanks in advance

**Comment by lreiher** [comment]
Services and actions are not supported, although it should in principle be pretty straightforward to implement. You might want to consider https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds.

**Reply by petergerten** [reply]
Any hints on what would need to be done to support services and actions? I think this would be very useful and only then makes it a full 'bridge' to ros2.

**Reply by lreiher** [reply]
I may have been a little too rash with my previous claim of *straightforward* implementation. ;)

Currently, the `mqtt_client` is using generic subscriptions and generic publishers for topics, [see here](https://github.com/ika-rwth-aachen/mqtt_client/blob/main/mqtt_client/src/MqttClient.ros2.cpp#L816). This allows us to work with arbitrary ROS message types at runtime.

Something similar would be needed for services and actions, however, I'm not aware of any `rclcpp`-functionality for generic services/actions. This is not to say that it's impossible, but would probably increase the implementation effort a lot.
